### PR TITLE
BIP-X: Morning Auction

### PR DESCRIPTION
**Pull request by BrianOlympus**
## Problem
From season to season, the temperature changes a maximum of +/-3% based on the state of Beanstalk. In instances where short-term demand exceeds short-term supply, Beanstalk may be paying significantly more than what is needed to attract creditors. While Beanstalk would rather have too high of a temperature rather than too low and risk not attracting creditors at all, the temperature can better optimize during times of excess demand. 

## Proposed Solution

Introduce a dutch auction during the first X blocks after the `sunrise()` call, where the Temperature ramps from 1% to 100% of the max Temperature. This time period will be referred as the `morning` within a season, and the mechanism named "the Morning Auction". (WIP)

Additionally, the soil issued above peg will change as a function of the Morning Auction, such that the maximum potential pods issued within a season (peas, WIP), will be reached if all soil is sown. 

## Specification

1. Redefine all instances of $h_t$ as the maximum Temperature $h_t^{\text{max}}$.

2. Redefine $h$ to:

$$ h =
\begin{cases}
{1} & if \quad c = 0 \\
{max(h_t^{\text{max}}*log_{ab+1}(ac+1),1)} & if \quad 0 < c  < 25 \\
{h_t^{\text{max}}} & else 
\end{cases} $$

where $a$ is the control variable, $b$ is the number of blocks needed to elapse to reach max Temperature, and $c$ is the number of blocks that has elapsed since the `sunrise()` call.

$$
\begin{aligned}
a =& \ 2 \\
b =& \ 25 \ (i.e \ \text{25 blocks until Temperature reaches its max)} \\
c =& \ \text{currentBlock} \ - \text{sunriseBlock}
\end{aligned}
$$

3. Adjust the time in which Beanstalk considers demand for Soil increasing in instances where all but at most one Soil is Sown from 5 minutes to 10 minutes (*i.e.*, $\Delta{E_{t-1}^{u^{first}}} \leq 600$).

### Storage
Add to the `Season` struct: `uint32 sunriseBlock`.
Add to the `Season` struct: `bool AbovePeg`.
Add to the `Field` struct: `uint128 beanSown`.
Modify `soil` in the `Season` struct from datatype `uint256` to `uint128` in favor of `beanSown`.
Depreciate `startSoil` from the Weather struct

### Existing Functions in scope
**SeasonFacet**
* `stepSeason()`

**LibDibbler**
* `sow(uint256 amount, address account)`
* `BeansToPods(uint256 beans, uint256 weather)`

**FieldFacet**
* `sow(uint256 amount, LibTransfer.From mode)` 
* `sowNoSoil(uint256 amount, address account)`
* `sowWithMin(uint256 amount, uint256 minAmount, LibTransfer.From mode)`
* `totalSoil()`

**weather**
* `stepWeather(int256 deltaB):`

#### New Functions/Getters: 
**LibDibbler**
* `MorningAuction(uint256 maxWeather, uint256 _block)` - outputs the new scaled weather as a function of the maximum weather and the blocks elapsed since sunrise.
* `AuctionMath(uint256 a)` - Math Helper for Morning Auction.
* `beansToPodsAbovePeg(uint256 beans, uint256 maxPeas)` - issue pods when bean is above peg. 

**FieldFacet**
* `yield()` -  outputs the instantaneous temperature in the season (max temp has a getter via `yield()`
* `peas()` - outputs Peas, or the amount of pods beanstalk is willing to issue. 

**SeasonFacet** 
* `abovePeg()` - getter for `bool abovePeg`
* `sunriseBlock()` - getter for `uint32 sunriseBlock`

## Technical Implementation: 

With this change, the temperature is dynamic within the season. 
When Bean is above peg, the soil issued within a season is dynamic, and the maximum pods (peas) issued within a season is fixed. 
When Bean is below peg, the soil issued within a season is fixed, and the peas is dynamic. 

image

Over the 25 blocks within a season, the temperature starts at 1% (absolute), and ramps up to 100% of the temperature. The formula can be seen here:  https://www.desmos.com/calculator/soynntzc7h 

`s.f.soil` and `totalSoil()` now behaves differently based on the peg. 
image
When above peg, beanstalk issues the effective soil needed if all  soil was sown immediately after sunrise (the pods harvested from previous season). when soil is sown, the amount sown is scaled up to reflect the effective soil consumed. 

## Execution
Estimated development time: **2-3 weeks** (lol).

1. Push proof-of-concept PR.
2. Beanstalk Farms code review.
3. External audit.
6. BIP.

### CODE REVIEW

File: protocol/contracts/libraries/LibIncentive.sol [hunk 1]
```diff
@@ -159,4 +162,14 @@ library LibIncentive {
         // 10**(36-decimals)
         return [1e30, C.curve3Pool().get_virtual_price()];
     }
+
+    function getEthUsdcPrice() private view returns (uint256) {
+        (int24 tick,) = OracleLibrary.consult(C.UniV3EthUsdc(),3600); //1 season tick
```
Comment (chaikitty) [inline protocol/contracts/libraries/LibIncentive.sol:167]: Did you mean to use PERIOD constant?

File: protocol/contracts/libraries/Token/LibTransfer.sol [hunk 2]
```diff
@@ -110,20 +110,4 @@ library LibTransfer {
             token.burn(burnt);
         }
     }
-
-
-    // ask publius, address(this) would refer to the beanstalk addy right
```
Comment (chaikitty) [inline protocol/contracts/libraries/Token/LibTransfer.sol]: I also removed this function in my branch, but I believe I removed the other one haha, lets check this when we re-merge

File: protocol/contracts/libraries/LibIncentive.sol (hunk 1 above)
Comment (BrianOlympus) [reply protocol/contracts/libraries/LibIncentive.sol:167]: yup, fixed!

**Comment by BrianOlympus** [comment]
closed in favor of BIP-30 PR: https://github.com/BeanstalkFarms/Beanstalk/pull/127

**Pull request closed without merging**